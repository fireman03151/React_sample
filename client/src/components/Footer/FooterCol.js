import React from 'react';
import PropTypes from 'prop-types';

import Link from '../helpers/Link';
import { dasherize } from '../../../../utils/slugs';

function FooterCol({ title, links }) {
  return (
    <div className={`footer-col ${dasherize(title)}`}>
      {title ? (
        <div className={`col-header `}>{title}</div>
      ) : (
        <div className='col-spacer'></div>
      )}
      {links.map(({ to, text, internal }, i) => (
        <Link external={!internal} key={`link-${i}`} to={to}>
          {text}
        </Link>
      ))}
    </div>
  );
}

const propTypes = {
  links: PropTypes.arrayOf(
    PropTypes.shape({
      to: PropTypes.string.isRequired,
      text: PropTypes.string,
      internal: PropTypes.bool
    })
  ).isRequired,
  title: PropTypes.string
};

FooterCol.propTypes = propTypes;
FooterCol.displayName = 'FooterCol';
export default FooterCol;
