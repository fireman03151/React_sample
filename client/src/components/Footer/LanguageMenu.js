/* eslint-disable jsx-a11y/no-onchange */
import React from 'react';
import { useTranslation } from 'react-i18next';

const {
  availableLangs,
  i18nextCodes,
  langDisplayNames
} = require('../../../i18n/allLangs');
const { homeLocation } = require('../../../config/env');

const locales = availableLangs.client;

const LanguageMenu = () => {
  const { i18n, t } = useTranslation();
  const i18nLanguage = i18n.language;

  const currentLanguage = Object.keys(i18nextCodes).find(
    key => i18nextCodes[key] === i18nLanguage
  );

  const changeLanguage = e => {
    const path = window.location.pathname;

    if (e.target.value === 'espanol') {
      window.location.replace(`${homeLocation}/espanol${path}`);
    } else {
      window.location.replace(`${homeLocation}${path}`);
    }
  };

  return (
    <div className='language-menu'>
      <label>
        {t('footer.language')}
        <select onChange={e => changeLanguage(e)} value={currentLanguage}>
          {locales.map((lang, i) => {
            return (
              <option key={i} value={lang}>
                {langDisplayNames[lang]}
              </option>
            );
          })}
        </select>
      </label>
    </div>
  );
};

export default LanguageMenu;
