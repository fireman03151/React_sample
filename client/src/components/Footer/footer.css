.footer {
  background-color: #f1f5f8;
  font-size: 0.9em;
  width: 100%;
  padding-top: 40px;
  margin-top: auto;
}

.footer .col-header {
  padding-bottom: 5px;
  padding-left: 5px;
  font-weight: bold;
}

.footer a:not([class*='inline']) {
  display: block;
  padding-left: 5px
}

.footer a:hover {
  outline-offset: 1px;
  outline: 1px dashed #A0A0A0;
  background-color: #ffbe41;
}

.night .footer {
  color: #606f7b;
  background: #222;
}

.night .footer a {
  color: #006400;
}

.night .footer a:hover {
  color: #606f7b;
  outline-offset: 1px;
  outline: 1px dashed #A0A0A0;
}
