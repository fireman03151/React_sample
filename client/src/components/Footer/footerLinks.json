{
  "footerLinks": [
    {
      "title": "Our Nonprofit",
      "links": [
        { "to": "/news/about/", "text": "About" },
        {
          "to": "https://www.linkedin.com/school/free-code-camp/people/",
          "text": "Alumni Network"
        },
        { "to": "https://github.com/freeCodeCamp/", "text": "Open Source" },
        { "to": "/news/shop/", "text": "Shop" },
        { "to": "/news/support/", "text": "Support" },
        { "to": "/news/sponsors/", "text": "Sponsors" },
        {
          "to": "/news/academic-honesty-policy/",
          "text": "Academic Honesty"
        },
        { "to": "/news/code-of-conduct/", "text": "Code of Conduct" },
        { "to": "/news/privacy-policy/", "text": "Privacy Policy" },
        { "to": "/news/terms-of-service/", "text": "Terms of Service" },
        { "to": "/news/copyright-policy/", "text": "Copyright Policy" }
      ]
    },
    {
      "title": "Trending Guides",
      "links": [
        {
          "to": "/news/2019-web-developer-roadmap/",
          "text": "2019 Web Developer Roadmap"
        },
        { "to": "/news/best-python-tutorial/", "text": "Python Tutorial" },
        {
          "to": "/news/understanding-flexbox-everything-you-need-to-know-b4013d4dc9af/",
          "text": "CSS Flexbox Guide"
        },
        {
          "to": "/news/best-javascript-tutorial/",
          "text": "JavaScript Tutorial"
        },
        { "to": "/news/python-example/", "text": "Python Example" },
        { "to": "/news/best-html-html5-tutorial/", "text": "HTML Tutorial" },
        {
          "to": "/news/linux-command-line-bash-tutorial/",
          "text": "Linux Command Line Guide"
        },
        { "to": "/news/javascript-example/", "text": "JavaScript Example" },

        { "to": "/news/best-git-tutorial/", "text": "Git Tutorial" },
        {
          "to": "/news/best-react-javascript-tutorial/",
          "text": "React Tutorial"
        },
        { "to": "/news/best-java-8-tutorial/", "text": "Java Tutorial" }
      ]
    },
    {
      "title": "",
      "links": [
        { "to": "/news/the-best-linux-tutorials/", "text": "Linux Tutorial" },

        { "to": "/news/best-css-and-css3-tutorial/", "text": "CSS Tutorial" },
        { "to": "/news/the-best-jquery-examples/", "text": "jQuery Example" },
        { "to": "/news/best-sql-database-tutorial/", "text": "SQL Tutorial" },
        { "to": "/news/css-example-css3/", "text": "CSS Example" },
        { "to": "/news/react-examples-reactjs/", "text": "React Example" },
        {
          "to": "/news/best-angular-tutorial-angularjs/",
          "text": "Angular Tutorial"
        },
        {
          "to": "/news/the-best-bootstrap-examples/",
          "text": "Bootstrap Example"
        },
        {
          "to": "/news/the-ultimate-guide-to-ssh-setting-up-ssh-keys/",
          "text": "How to Set Up SSH Keys"
        },
        {
          "to": "/news/best-wordpress-tutorial/",
          "text": "WordPress Tutorial"
        },
        { "to": "/news/the-best-php-examples/", "text": "PHP Example" }
      ]
    }
  ]
}
