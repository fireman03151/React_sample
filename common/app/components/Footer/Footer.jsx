import React from 'react';
import { Col, Row, Grid } from 'react-bootstrap';

import links from './links.json';

export default class extends React.Component {
  static displayName = 'Footer';
  renderLinks(mobile) {
    return links.map(link => {
      return (
        <a
          className={ link.className}
          href={ link.href }
          key={ link.content }
          target={ link.target }>
          { this.renderContent(mobile, link.content) }
        </a>
      );
    });
  }

  renderContent(mobile, content) {
    if (mobile) {
      return (
        <span className='sr-only'>
          content;
        </span>
      );
    }
    return content;
  }

  render() {
    return (
      <Grid className='fcc-footer'>
        <Row>
          <Col
            className='hidden-xs hidden-sm'
            xs={ 12 }>
            { this.renderLinks() }
          </Col>
          <Col
            className='visible-xs visible-sm'
            xs={ 12 }>
            { this.renderLinks(true) }
          </Col>
        </Row>
      </Grid>
    );
  }
}
