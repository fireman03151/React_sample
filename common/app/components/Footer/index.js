export { default as Footer } from './Footer.jsx';
