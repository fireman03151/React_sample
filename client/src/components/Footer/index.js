import React from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import Link from '../helpers/Link';
import LanguageMenu from './LanguageMenu';
import './footer.css';

const { showLocaleDropdownMenu = false } = require('../../../config/env');

const propTypes = {
  children: PropTypes.any
};

const ColHeader = ({ children, ...other }) => (
  <div className='col-header' {...other}>
    {children}
  </div>
);
ColHeader.propTypes = propTypes;

function Footer() {
  const { t } = useTranslation();

  return (
    <footer className='site-footer'>
      <div className='footer-container'>
        <div className='footer-top'>
          <div className='footer-desc-col'>
            {showLocaleDropdownMenu ? <LanguageMenu /> : null}
            <p>{t('footer.tax-exempt-status')}</p>
            <p>{t('footer.mission-statement')}</p>
            <p>{t('footer.donation-initiatives')}</p>
            <p className='footer-donation'>
              {t('footer.donate-text')}{' '}
              <Link className='inline' to='/donate'>
                {t('footer.donate-link')}
              </Link>
              .
            </p>
          </div>
          <div className='trending-guides'>
            <div className='col-header'>{t('footer.trending-guides')}</div>
            <div className='trending-guides-row'>
              <div className='footer-col footer-col-1'>
                <Link external={false} to={t('trending:article1link')}>
                  {t('trending:article1title')}
                </Link>
                <Link external={false} to={t('trending:article2link')}>
                  {t('trending:article2title')}
                </Link>
                <Link external={false} to={t('trending:article3link')}>
                  {t('trending:article3title')}
                </Link>
                <Link external={false} to={t('trending:article4link')}>
                  {t('trending:article4title')}
                </Link>
                <Link external={false} to={t('trending:article5link')}>
                  {t('trending:article5title')}
                </Link>
                <Link external={false} to={t('trending:article6link')}>
                  {t('trending:article6title')}
                </Link>
                <Link external={false} to={t('trending:article7link')}>
                  {t('trending:article7title')}
                </Link>
                <Link external={false} to={t('trending:article8link')}>
                  {t('trending:article8title')}
                </Link>
                <Link external={false} to={t('trending:article9link')}>
                  {t('trending:article9title')}
                </Link>
                <Link external={false} to={t('trending:article10link')}>
                  {t('trending:article10title')}
                </Link>
              </div>
              <div className='footer-col footer-col-2'>
                <Link external={false} to={t('trending:article11link')}>
                  {t('trending:article11title')}
                </Link>
                <Link external={false} to={t('trending:article12link')}>
                  {t('trending:article12title')}
                </Link>
                <Link external={false} to={t('trending:article13link')}>
                  {t('trending:article13title')}
                </Link>
                <Link external={false} to={t('trending:article14link')}>
                  {t('trending:article14title')}
                </Link>
                <Link external={false} to={t('trending:article15link')}>
                  {t('trending:article15title')}
                </Link>
                <Link external={false} to={t('trending:article16link')}>
                  {t('trending:article16title')}
                </Link>
                <Link external={false} to={t('trending:article17link')}>
                  {t('trending:article17title')}
                </Link>
                <Link external={false} to={t('trending:article18link')}>
                  {t('trending:article18title')}
                </Link>
                <Link external={false} to={t('trending:article19link')}>
                  {t('trending:article19title')}
                </Link>
                <Link external={false} to={t('trending:article20link')}>
                  {t('trending:article20title')}
                </Link>
              </div>
              <div className='footer-col footer-col-3'>
                <div className='footer-left'>
                  <Link external={false} to={t('trending:article21link')}>
                    {t('trending:article21title')}
                  </Link>
                  <Link external={false} to={t('trending:article22link')}>
                    {t('trending:article22title')}
                  </Link>
                  <Link external={false} to={t('trending:article23link')}>
                    {t('trending:article23title')}
                  </Link>
                  <Link external={false} to={t('trending:article24link')}>
                    {t('trending:article24title')}
                  </Link>
                  <Link external={false} to={t('trending:article25link')}>
                    {t('trending:article25title')}
                  </Link>
                </div>

                <div className='footer-right'>
                  <Link external={false} to={t('trending:article26link')}>
                    {t('trending:article26title')}
                  </Link>
                  <Link external={false} to={t('trending:article27link')}>
                    {t('trending:article27title')}
                  </Link>
                  <Link external={false} to={t('trending:article28link')}>
                    {t('trending:article28title')}
                  </Link>
                  <Link external={false} to={t('trending:article29link')}>
                    {t('trending:article29title')}
                  </Link>
                  <Link external={false} to={t('trending:article30link')}>
                    {t('trending:article30title')}
                  </Link>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div className='footer-bottom'>
          <div className='col-header'>{t('footer.our-nonprofit')}</div>
          <div className='footer-divder' />
          <div className='our-nonprofit'>
            <Link
              external={false}
              to={'https://www.freecodecamp.org/news/about/'}
            >
              {t('footer.links.about')}
            </Link>
            <Link
              external={false}
              sameTab={false}
              to={'https://www.linkedin.com/school/free-code-camp/people/'}
            >
              {t('footer.links.alumni')}
            </Link>
            <Link external={false} to={'https://github.com/freeCodeCamp/'}>
              {t('footer.links.open-source')}
            </Link>
            <Link
              external={false}
              sameTab={false}
              to={'https://www.freecodecamp.org/shop/'}
            >
              {t('footer.links.shop')}
            </Link>
            <Link
              external={false}
              to={'https://www.freecodecamp.org/news/support/'}
            >
              {t('footer.links.support')}
            </Link>
            <Link
              external={false}
              to={'https://www.freecodecamp.org/news/sponsors/'}
            >
              {t('footer.links.sponsors')}
            </Link>
            <Link
              external={false}
              to={'https://www.freecodecamp.org/news/academic-honesty-policy/'}
            >
              {t('footer.links.honesty')}
            </Link>
            <Link
              external={false}
              to={'https://www.freecodecamp.org/news/code-of-conduct/'}
            >
              {t('footer.links.coc')}
            </Link>
            <Link
              external={false}
              to={'https://www.freecodecamp.org/news/privacy-policy/'}
            >
              {t('footer.links.privacy')}
            </Link>
            <Link
              external={false}
              to={'https://www.freecodecamp.org/news/terms-of-service/'}
            >
              {t('footer.links.tos')}
            </Link>
            <Link
              external={false}
              to={'https://www.freecodecamp.org/news/copyright-policy/'}
            >
              {t('footer.links.copyright')}
            </Link>
          </div>
        </div>
      </div>
    </footer>
  );
}

Footer.displayName = 'Footer';
export default Footer;
