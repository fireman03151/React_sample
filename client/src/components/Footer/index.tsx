import React from 'react';
import { Trans, useTranslation } from 'react-i18next';
import Link from '../helpers/link';
import './footer.css';

function Footer(): JSX.Element {
  const { t } = useTranslation();

  return (
    <footer className='site-footer'>
      <div className='footer-top'>
        <div className='footer-desc-col'>
          <p>{t('footer.tax-exempt-status')}</p>
          <p>{t('footer.mission-statement')}</p>
          <p>{t('footer.donation-initiatives')}</p>
          <p className='footer-donation'>
            <Trans i18nKey='footer.donate-text'>
              You can
              <Link className='inline' to='/donate'>
                make a tax-deductible donation here
              </Link>
              .
            </Trans>
          </p>
        </div>
        <div className='trending-guides'>
          <h2 className='col-header'>{t('footer.trending-guides')}</h2>
          <div className='trending-guides-articles'>
            {/* the span elements were add so the link can be inline and its text would be the only active target */}
            <span>
              <Link external={false} to={t('trending:article0link')}>
                {t('trending:article0title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article1link')}>
                {t('trending:article1title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article2link')}>
                {t('trending:article2title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article3link')}>
                {t('trending:article3title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article4link')}>
                {t('trending:article4title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article5link')}>
                {t('trending:article5title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article6link')}>
                {t('trending:article6title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article7link')}>
                {t('trending:article7title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article8link')}>
                {t('trending:article8title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article9link')}>
                {t('trending:article9title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article10link')}>
                {t('trending:article10title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article11link')}>
                {t('trending:article11title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article12link')}>
                {t('trending:article12title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article13link')}>
                {t('trending:article13title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article14link')}>
                {t('trending:article14title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article15link')}>
                {t('trending:article15title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article16link')}>
                {t('trending:article16title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article17link')}>
                {t('trending:article17title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article18link')}>
                {t('trending:article18title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article19link')}>
                {t('trending:article19title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article20link')}>
                {t('trending:article20title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article21link')}>
                {t('trending:article21title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article22link')}>
                {t('trending:article22title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article23link')}>
                {t('trending:article23title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article24link')}>
                {t('trending:article24title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article25link')}>
                {t('trending:article25title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article26link')}>
                {t('trending:article26title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article27link')}>
                {t('trending:article27title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article28link')}>
                {t('trending:article28title')}
              </Link>
            </span>
            <span>
              <Link external={false} to={t('trending:article29link')}>
                {t('trending:article29title')}
              </Link>
            </span>
          </div>
        </div>
      </div>
      <div className='footer-bottom'>
        <h2 className='col-header'>{t('footer.our-nonprofit')}</h2>
        <div className='our-nonprofit'>
          <Link external={false} to={t('links:footer.about-url')}>
            {t('footer.links.about')}
          </Link>
          <Link
            external={false}
            sameTab={false}
            to={'https://www.linkedin.com/school/free-code-camp/people/'}
          >
            {t('footer.links.alumni')}
          </Link>
          <Link external={false} to={'https://github.com/freeCodeCamp/'}>
            {t('footer.links.open-source')}
          </Link>
          <Link
            external={false}
            sameTab={false}
            to={t('links:footer.shop-url')}
          >
            {t('footer.links.shop')}
          </Link>
          <Link external={false} to={t('links:footer.support-url')}>
            {t('footer.links.support')}
          </Link>
          <Link external={false} to={t('links:footer.sponsors-url')}>
            {t('footer.links.sponsors')}
          </Link>
          <Link external={false} to={t('links:footer.honesty-url')}>
            {t('footer.links.honesty')}
          </Link>
          <Link external={false} to={t('links:footer.coc-url')}>
            {t('footer.links.coc')}
          </Link>
          <Link external={false} to={t('links:footer.privacy-url')}>
            {t('footer.links.privacy')}
          </Link>
          <Link external={false} to={t('links:footer.tos-url')}>
            {t('footer.links.tos')}
          </Link>
          <Link external={false} to={t('links:footer.copyright-url')}>
            {t('footer.links.copyright')}
          </Link>
        </div>
      </div>
    </footer>
  );
}

Footer.displayName = 'Footer';
export default Footer;
