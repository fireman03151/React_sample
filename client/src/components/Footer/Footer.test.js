/* global expect */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-dom/extend-expect';

import Footer from './';

describe('<Footer />', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(<Footer />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
