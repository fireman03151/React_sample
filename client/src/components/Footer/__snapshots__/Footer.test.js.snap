// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> matches snapshot 1`] = `
<footer
  className="site-footer"
>
  <div
    className="footer-container"
  >
    <div
      className="footer-top"
    >
      <div
        className="footer-desc-col"
      >
        <div
          className="language-menu"
        >
          <span>
            footer.language
          </span>
          <select
            onChange={[Function]}
          >
            <option
              value="english"
            >
              English
            </option>
            <option
              value="espanol"
            >
              Español
            </option>
            <option
              value="chinese"
            >
              中文
            </option>
          </select>
        </div>
        <p>
          footer.tax-exempt-status
        </p>
        <p>
          footer.mission-statement
        </p>
        <p>
          footer.donation-initiatives
        </p>
        <p
          className="footer-donation"
        >
          footer.donate-text
           
          <a
            className="inline"
            href="/donate"
          >
            footer.donate-link
          </a>
          .
        </p>
      </div>
      <div
        className="trending-guides"
      >
        <div
          className="col-header"
        >
          footer.trending-guides
        </div>
        <div
          className="trending-guides-row"
        >
          <div
            className="footer-col footer-col-1"
          >
            <a
              href="https://www.freecodecamp.org/news/git-clone-branch-how-to-clone-a-specific-branch/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Git Clone
            </a>
            <a
              href="https://www.freecodecamp.org/news/agile-methods-and-methodology-for-beginners/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Agile Methods
            </a>
            <a
              href="https://www.freecodecamp.org/news/if-name-main-python-example/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Python Main
            </a>
            <a
              href="https://www.freecodecamp.org/news/javascript-callback-functions-what-are-callbacks-in-js-and-how-to-use-them/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Callback
            </a>
            <a
              href="https://www.freecodecamp.org/news/debounce-javascript-tutorial-how-to-make-your-js-wait-up/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Debounce
            </a>
            <a
              href="https://www.freecodecamp.org/news/javascript-url-encode-example-how-to-use-encodeuricomponent-and-encodeuri/"
              rel="noopener noreferrer"
              target="_blank"
            >
              URL Encode
            </a>
            <a
              href="https://www.freecodecamp.org/news/make-it-blink-html-tutorial-how-to-use-the-blink-tag-with-code-examples/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Blink HTML
            </a>
            <a
              href="https://www.freecodecamp.org/news/python-returns-multiple-values-how-to-return-a-tuple-list-dictionary/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Python Tuple
            </a>
            <a
              href="https://www.freecodecamp.org/news/javascript-array-insert-how-to-add-to-an-array-with-the-push-unshift-and-concat-functions/"
              rel="noopener noreferrer"
              target="_blank"
            >
              JavaScript Push
            </a>
            <a
              href="https://www.freecodecamp.org/news/java-list-tutorial-util-list-api-example/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Java List
            </a>
          </div>
          <div
            className="footer-col footer-col-2"
          >
            <a
              href="https://www.freecodecamp.org/news/learn-ux-design-self-taught-user-experience-designer/"
              rel="noopener noreferrer"
              target="_blank"
            >
              UX
            </a>
            <a
              href="https://www.freecodecamp.org/news/what-is-design-thinking-an-introduction-to-the-design-process-for-entrepreneurs-and-developers/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Design Thinking
            </a>
            <a
              href="https://www.freecodecamp.org/news/prime-numbers-list-chart-of-primes/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Prime Number List
            </a>
            <a
              href="https://www.freecodecamp.org/news/product-design-explained-in-plain-english/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Product Design
            </a>
            <a
              href="https://www.freecodecamp.org/news/what-is-digital-design-and-why-does-it-matter/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Digital Design
            </a>
            <a
              href="https://www.freecodecamp.org/news/best-coding-games-online-adults-learn-to-code/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Coding Games
            </a>
            <a
              href="https://www.freecodecamp.org/news/svm-machine-learning-tutorial-what-is-the-support-vector-machine-algorithm-explained-with-code-examples/"
              rel="noopener noreferrer"
              target="_blank"
            >
              SVM
            </a>
            <a
              href="https://www.freecodecamp.org/news/javascript-foreach-how-to-loop-through-an-array-in-js/"
              rel="noopener noreferrer"
              target="_blank"
            >
              JavaScript forEach
            </a>
            <a
              href="https://www.freecodecamp.org/news/google-bert-nlp-machine-learning-tutorial/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Google BERT
            </a>
            <a
              href="https://www.freecodecamp.org/news/sql-create-table-statement-with-example-syntax/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Create Table SQL
            </a>
          </div>
          <div
            className="footer-col footer-col-3"
          >
            <div
              className="footer-left"
            >
              <a
                href="https://www.freecodecamp.org/news/responsive-web-design-how-to-make-a-website-look-good-on-phones-and-tablets/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Responsive Web Design
              </a>
              <a
                href="https://www.freecodecamp.org/news/svg-basics-what-are-scalable-vector-graphics-and-how-do-you-use-them/"
                rel="noopener noreferrer"
                target="_blank"
              >
                What Is an SVG File?
              </a>
              <a
                href="https://www.freecodecamp.org/news/pdf-password-remover-guide-how-to-remove-password-protection-from-a-pdf/"
                rel="noopener noreferrer"
                target="_blank"
              >
                PDF Password Remover
              </a>
              <a
                href="https://www.freecodecamp.org/news/what-is-a-pdf-file-and-how-do-you-open-it-solved/"
                rel="noopener noreferrer"
                target="_blank"
              >
                What Is a PDF?
              </a>
              <a
                href="https://www.freecodecamp.org/news/what-is-python-used-for-10-coding-uses-for-the-python-programming-language/"
                rel="noopener noreferrer"
                target="_blank"
              >
                What Is Python?
              </a>
            </div>
            <div
              className="footer-right"
            >
              <a
                href="https://www.freecodecamp.org/news/what-is-tls-transport-layer-security-encryption-explained-in-plain-english/"
                rel="noopener noreferrer"
                target="_blank"
              >
                What Is TLS?
              </a>
              <a
                href="https://www.freecodecamp.org/news/what-is-a-lan-local-area-network-explained-in-plain-english/"
                rel="noopener noreferrer"
                target="_blank"
              >
                What Is a LAN?
              </a>
              <a
                href="https://www.freecodecamp.org/news/what-is-npm-a-node-package-manager-tutorial-for-beginners/"
                rel="noopener noreferrer"
                target="_blank"
              >
                What Is npm?
              </a>
              <a
                href="https://www.freecodecamp.org/news/rsync-examples-rsync-options-and-how-to-copy-files-over-ssh/"
                rel="noopener noreferrer"
                target="_blank"
              >
                RSync Examples
              </a>
              <a
                href="https://www.freecodecamp.org/news/how-to-use-the-tree-based-algorithm-for-machine-learning/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Random Forest
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="footer-buttom"
    >
      <div
        className="col-header"
      >
        footer.our-nonprofit
      </div>
      <div
        className="footer-divder"
      />
      <div
        className="our-nonprofit"
      >
        <a
          href="https://www.freecodecamp.org/news/about/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.about
        </a>
        <a
          href="https://www.linkedin.com/school/free-code-camp/people/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.alumni
        </a>
        <a
          href="https://github.com/freeCodeCamp/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.open-source
        </a>
        <a
          href="https://www.freecodecamp.org/shop/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.shop
        </a>
        <a
          href="https://www.freecodecamp.org/news/support/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.support
        </a>
        <a
          href="https://www.freecodecamp.org/news/sponsors/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.sponsors
        </a>
        <a
          href="https://www.freecodecamp.org/news/academic-honesty-policy/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.honesty
        </a>
        <a
          href="https://www.freecodecamp.org/news/code-of-conduct/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.coc
        </a>
        <a
          href="https://www.freecodecamp.org/news/privacy-policy/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.privacy
        </a>
        <a
          href="https://www.freecodecamp.org/news/terms-of-service/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.tos
        </a>
        <a
          href="https://www.freecodecamp.org/news/copyright-policy/"
          rel="noopener noreferrer"
          target="_blank"
        >
          footer.links.copyright
        </a>
      </div>
    </div>
  </div>
</footer>
`;
